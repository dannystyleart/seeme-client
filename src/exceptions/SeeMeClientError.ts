export class SeeMeClientError extends Error {
  constructor(message?: string) {
    super(`SeeMeClientError: ${message}`);
  }
}
