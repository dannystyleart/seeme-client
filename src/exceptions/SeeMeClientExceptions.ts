export enum SeeMeClientExceptions {
  API_KEY_MISSING = 'API_KEY_MISSING',
  API_HOST_MISSING = 'API_HOST_MISSING',
  API_HOST_PROTOCOL_UNSUPPORTED = 'API_HOST_PROTOCOL_UNSUPPORTED',
  API_PATH_INVALID = 'API_PATH_INVALID',
  SETIP_INVALID_IP = 'SETIP_INVALID_IP'
}
