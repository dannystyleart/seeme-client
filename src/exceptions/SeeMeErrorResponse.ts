export class SeeMeErrorResponse extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'SeeMeErrorResponse';
  }
}
