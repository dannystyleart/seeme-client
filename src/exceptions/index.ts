export * from './SeeMeClientError';
export * from './SeeMeErrorResponse';
export * from './SeeMeClientExceptions';
