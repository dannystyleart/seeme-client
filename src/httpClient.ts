import { URLSearchParams } from 'url';
import fetch from 'node-fetch';
import { removeSlash } from './utils';

export const createApiClient = (apiUrl: string) => {
  return <T>(path: string, params: any): Promise<T> => {
    const apiEndpoint = [removeSlash(apiUrl), removeSlash(path, false)].join('/');
    const queryParams = new URLSearchParams(<any>params);

    return new Promise(async (resolve, reject) => {
      try {
        const response = await fetch([apiEndpoint, queryParams.toString()].join('?'));
        resolve(response.json());
      } catch (e) {
        reject(e);
      }
    });
  };
};
