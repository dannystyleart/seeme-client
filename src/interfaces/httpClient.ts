export interface HttpClientRequest {
  (path: string, params?: any): Promise<any>;
}
