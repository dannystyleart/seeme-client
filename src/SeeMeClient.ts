import { isHttpProtocol, isNonEmptyString, withDefaults } from './utils';
import { SeeMeClient, SeeMeClientOptions } from './interfaces';
import { SeeMeClientError, SeeMeClientExceptions } from './exceptions';
import { createApiClient } from './httpClient';

const defaultConfig = {
  apiHost: 'https://seeme.hu',
  apiPath: '/gateway',
  apiVersion: '2.0.1'
};

export const createClient = (options: SeeMeClientOptions): SeeMeClient => {
  const clientOptions = withDefaults(options, defaultConfig) as SeeMeClientOptions;
  const { apiKey, apiHost, apiPath, apiVersion } = clientOptions;

  if (!isNonEmptyString(apiKey)) throw new SeeMeClientError(SeeMeClientExceptions.API_KEY_MISSING);
  if (!isNonEmptyString(apiHost)) throw new SeeMeClientError(SeeMeClientExceptions.API_HOST_MISSING);
  if (!isHttpProtocol(apiHost)) throw new SeeMeClientError(SeeMeClientExceptions.API_HOST_PROTOCOL_UNSUPPORTED);
  if (!isNonEmptyString(apiPath)) throw new SeeMeClientError(SeeMeClientExceptions.API_PATH_INVALID);

  const request = createApiClient(apiHost);

  const gatewayCall = (params: any = {}) =>
    request(apiPath, {
      ...params,
      key: apiKey,
      format: 'json',
      apiVersion
    });

  const notImplemented = () => Promise.reject('NOT_YET_IMPLEMENTED');
  return {
    getBalance: () => gatewayCall({ method: 'balance' }),
    setIP: (ip) => gatewayCall({ method: 'setip', ip }),
    sendMessage: notImplemented
  };
};
