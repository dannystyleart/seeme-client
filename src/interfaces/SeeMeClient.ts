import { SendMessageCallbackStatus } from '../SeeMeClient';

export interface SeeMeClient {
  getBalance(): Promise<any>;
  setIP(ipAddress: string): Promise<any>;
  sendSMS(number: string, message: string, optionalParams?: SendMessageOptionalParams): Promise<any>;
}

export interface SendMessageOptionalParams {
  sender?: string;
  referenceId?: string | number;
  callbackStatus?: 'all' | SendMessageCallbackStatus | Array<SendMessageCallbackStatus>;
  callbackURL?: string;
}
