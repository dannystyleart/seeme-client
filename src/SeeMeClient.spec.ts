import * as httpClient from './httpClient';
import { createClient } from './SeeMeClient';
import { SeeMeClientError, SeeMeClientExceptions } from './exceptions';

jest.mock('./httpClient');
describe('SeeMeClient', () => {
  let httpApiClientFactory: jest.MockedFunction<any>;
  let apiClient: jest.MockedFunction<any>;
  beforeAll(() => {
    httpApiClientFactory = jest.fn();
    (httpClient.createApiClient as jest.MockedFunction<any>).mockImplementation(httpApiClientFactory);
  });

  beforeEach(() => {
    apiClient = jest.fn();
    apiClient.mockResolvedValue({});

    httpApiClientFactory.mockReturnValue(apiClient);
  });

  afterEach(() => {
    httpApiClientFactory.mockReset();
  });

  describe('creating client should', () => {
    test(`throw ${SeeMeClientExceptions.API_KEY_MISSING} error when option 'apiKey' not provided`, () => {
      const expectedError = new SeeMeClientError(SeeMeClientExceptions.API_KEY_MISSING);
      expect(() => createClient({} as any)).toThrowError(expectedError);
    });

    test(`throw ${SeeMeClientExceptions.API_HOST_MISSING} error when provided 'apiHost' option with empty value`, () => {
      const expectedError = new SeeMeClientError(SeeMeClientExceptions.API_HOST_MISSING);
      const options = {
        apiKey: '<api-key>',
        apiHost: ''
      };

      expect(() => createClient(options)).toThrowError(expectedError);
    });

    test(`throw ${SeeMeClientExceptions.API_HOST_PROTOCOL_UNSUPPORTED} error when provided 'apiHost' option with non http url`, () => {
      const expectedError = new SeeMeClientError(SeeMeClientExceptions.API_HOST_PROTOCOL_UNSUPPORTED);
      const options = {
        apiKey: '<api-key>',
        apiHost: 'ws://example.com:3000'
      };

      expect(() => createClient(options)).toThrowError(expectedError);
    });

    test(`throw ${SeeMeClientExceptions.API_PATH_INVALID} error when provided 'apiPath' option with empty value`, () => {
      const expectedError = new SeeMeClientError(SeeMeClientExceptions.API_PATH_INVALID);
      const options = {
        apiKey: '<api-key>',
        apiPath: ''
      };

      expect(() => createClient(options)).toThrowError(expectedError);
    });

    test('return configured client', () => {
      const options = {
        apiKey: '<api-key>',
        apiHost: 'http://example.com'
      };

      const result = createClient(options);

      expect(httpApiClientFactory).toHaveBeenCalledWith(options.apiHost);
      expect(result).toMatchSnapshot();
    });
  });

  describe('getBalance() should', () => {
    test('do api request correctly', async () => {
      const options = {
        apiKey: '<api-key>',
        apiHost: 'http://example.com',
        apiPath: '/endpoint'
      };
      const expectedApiClientParams = [
        options.apiPath,
        expect.objectContaining({
          method: 'balance',
          key: options.apiKey
        })
      ];
      await createClient(options).getBalance();

      expect(apiClient).toHaveBeenCalledWith(...expectedApiClientParams);
    });
  });

  describe('setIP(ip) should', () => {
    test(`throw ${SeeMeClientExceptions.SETIP_INVALID_IP} error for invalid ip param`, () => {
      const client = createClient({
        apiKey: '<api-key>'
      });

      [false, 1, '', 'random text'].forEach((ip) => {
        const probe = () => client.setIP(ip as any);
        const expectedError = new SeeMeClientError(SeeMeClientExceptions.SETIP_INVALID_IP);

        expect(probe).rejects.toThrowError(expectedError);
      });
    });

    test('do api request correctly', async () => {
      const testIp = '127.0.0.1';
      const options = {
        apiKey: '<api-key>',
        apiHost: 'http://example.com',
        apiPath: '/endpoint'
      };
      const expectedApiClientParams = [
        options.apiPath,
        expect.objectContaining({
          method: 'setip',
          ip: testIp,
          key: options.apiKey
        })
      ];

      await createClient(options).setIP(testIp);

      expect(apiClient).toHaveBeenCalledWith(...expectedApiClientParams);
    });
  });
});
