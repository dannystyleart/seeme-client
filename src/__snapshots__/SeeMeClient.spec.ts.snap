// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SeeMeClient creating client should return configured client 1`] = `
Object {
  "getBalance": [Function],
  "sendSMS": [Function],
  "setIP": [Function],
}
`;
